### ISSUE TITLE
CC: Add image request timeout configuration

### ISSUE THREAD

**Issue by stevenhorsman** [kata-containers/kata-containers#8141]
Once we have the ability to support image pulling on the guest (https://github.com/kata-containers/kata-containers/issues/8103), one of the features we can add to improve this, is the ability to set a image request timeout so that request that need to pull large images on the guest don't timeout.

We have a commit https://github.com/kata-containers/kata-containers/pull/7390/commits/57b932c1276ed83bc3a5bb61f005302130742433 that does this behaviour, but I think it's currently tied to the pull image API on the kata-agent, which with the nydsu-snapshotter we are removing, and this behaviour will be integrated into the createContainers request. As a result I think this might need updating to fit it. @ChengyuZhu6 is probably in the best place to confirm this as they have worked on both pieces.

I think it's worth adding a new e2e test to check this and we've added a very similar one in the cloud-api-adaptor (https://github.com/confidential-containers/cloud-api-adaptor/issues/1111) which included the dockerfile for building a large image, so we might want to adopt that and build it as part of our workflow as well

As part of this change, whilst we are updating the [image section](https://github.com/kata-containers/kata-containers/blob/fa6786d1d79f915c4b0985af57e174dd36641f39/src/runtime/config/configuration-qemu.toml.in#L695-L720) of the configuration file I think we should take the opportunity to clean up the values that we aren't using, which I think might even be all of them once we get to `main`?

## Acceptance Criteria

For each supported platform (TEEs and kubernetes & crictl)

_**Scenario: Image pull on guest of a large image works on Kubernetes**_
**Given** I have a version of kata deployed on Kubernetes with a guest image that has an agent with `guest_pull` feature enabled and nydus-snapshotter installed and configured for [guest-pulling](https://github.com/containerd/nydus-snapshotter/blob/main/misc/snapshotter/config-coco-guest-pulling.toml)
**And** I have configured the imageRequestTimeout to be larger than 60s e.g. 300s
**When** I try and create a pod from a large container image, which takes longer than 60s to be pulled
**Then** The pod is successfully created

**Comment by fitzthum** [comment]
Regarding the `image` section, I made a PR about this a few weeks ago.

**Comment by ariel-adam** [comment]
@fitzthum @stevenhorsman what should be the next steps on this issue? Who should own it?

**Comment by stevenhorsman** [comment]
This is blocked on image pull in the guest

**Comment by ChengyuZhu6** [comment]
If the task is currently inactive, I'll take it on.

**Comment by ChengyuZhu6** [comment]
It’s important to note that the image pulled in the guest is stored in memory, which we need adjust the memory allocated to the guest for running the test, such as modifying the `default_memory`. However, it has been observed that regardless of whether it’s a TEE or non-TDD environment, Kata Containers allocates only about one-tenth of `default_memory` to the /run directory by default. This presents a challenge when dealing with large images. For instance, with a `default_memory` setting of 2048 MB, the maximum image size that can be downloaded within the guest is limited to 204.8 MB. Simply increasing `default_memory` would lead to inefficient use of CI machine resources. To illustrate, testing a 4GB image would necessitate at least 40GB of memory, resulting in 36GB of unused memory.

**Comment by ChengyuZhu6** [comment]
Currently, there are two potential solutions to address this issue:

Adjust the ratio of `default_memory` to the memory allocated for the /run directory in the guest rootfs?
Directly set the memory size for the /run directory, similar to how `default_memory` is configured?

**Comment by ChengyuZhu6** [comment]
I'd like to hear from the experts on it.

**Comment by ChengyuZhu6** [comment]
Friendly ping @stevenhorsman @fitzthum @fidencio

**Comment by stevenhorsman** [comment]
Hi Chengyu, a couple of comments on this about the memory limitations. We have two main scenarios for image-pull on guest:
- local hypervsior - in `CCv0` Arron added the ability to pull to attached, trusted ephemeral storage. This has not been ported yet, but we have https://github.com/kata-containers/kata-containers/issues/8142, so I think this solves your concerns
- remote hypervisor (aka peer pods) - As we are using encrypted filesystems in the confdiential VMs we create as peer pods, we don't need to restrict ourselves to tmpfs in peer pods, so as such use system to re-mount [/run/image](https://github.com/confidential-containers/cloud-api-adaptor/blob/main/src/cloud-api-adaptor/podvm/files/etc/systemd/system/run-image.mount) and [/run/kata-containers](https://github.com/confidential-containers/cloud-api-adaptor/blob/main/src/cloud-api-adaptor/podvm/files/etc/systemd/system/run-kata%5Cx2dcontainers.mount) outside of tmpfs.

Based on these two things I think we have possible solutions to the memory required by large images, but still hit the hurdle of the time taken to request this large images being trouble, so that's why I scoped this issue to just deal with that via configuration.

I hope that helps explain the thoughts behind the image description, but as I noted my expectations might be outdated now, so please being up any challenges you see.

**Comment by ChengyuZhu6** [comment]
> Based on these two things I think we have possible solutions to the memory required by large images, but still hit the hurdle of the time taken to request this large images being trouble, so that's why I scoped this issue to just deal with that via configuration.

Thanks for your information @stevenhorsman. I agree that your suggestions could address my concerns. However, I’m uncertain about how to write the case to test large images in Kubernetes before the trusted storage code is merged into main.

Should we simply increate `default_memory`? Alternatively, should we write the test and mark it to be skipped until the trusted storage code is integrated? What are your thoughts?

If there is a more reasonable alternative to the two options mentioned, or if selecting one of them is preferable, I'm prepared to make some changes in the PR (kata-containers/kata-containers#9332), ensuring it is ready for review by next week.

**Comment by fitzthum** [comment]
One thing to note is that this 10 percent limitation arises from systemd. When using a rootfs without systemd `/run` will be mounted with 50 percent of the guest memory. This inconsistency is a bit weird. There was an [unsuccessful attempt](https://github.com/kata-containers/kata-containers/pull/7016) to fix this a while ago. Maybe we should look into fixing now. Doing so would help with large images significantly. 

In the long term I had been assuming that host-pulling would be the main way we would support really big images, but maybe that's a bad assumption.

**Comment by stevenhorsman** [comment]
> Should we simply increate default_memory? Alternatively, should we write the test and mark it to be skipped until the trusted storage code is integrated? What are your thoughts?

I think that's probably the most pragmatic approach. We can potentially change default memory manually in local test cases and potentially test this out on peer pods once we've switched over to main there before trusted storage is ready (we already have a test case for this on `CCv0`)

**Comment by ChengyuZhu6** [comment]
> One thing to note is that this 10 percent limitation arises from systemd. When using a rootfs without systemd `/run` will be mounted with 50 percent of the guest memory. This inconsistency is a bit weird. There was an [unsuccessful attempt](https://github.com/kata-containers/kata-containers/pull/7016) to fix this a while ago. Maybe we should look into fixing now. Doing so would help with large images significantly.
> 
> In the long term I had been assuming that host-pulling would be the main way we would support really big images, but maybe that's a bad assumption.

Thanks @fitzthum. That's what I tried to resolve the issue on my local machine.
